internal/weldrcheck/blueprints_test.go

```diff
@@ -0,0 +1,754 @@
+// The blueprint version number may get bumped if the server has had tests run before
+// do not assume the bp version will match unless first deleting the old one.
+
+// +build integration
+
+package weldrcheck
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	//	"strings"
+	"testing"
+	"time"
+
+	"github.com/BurntSushi/toml"
+	"github.com/stretchr/testify/require"
+
+	"github.com/osbuild/osbuild-composer/internal/client"
+)
+
+// Hold test state to share between tests
+var testState *TestState
+
+// Setup the socket to use for running the tests
+// Also makes sure there is a running server to test against
+func TestMain(m *testing.M) {
+	var err error
+	testState, err = setupTestState("/run/weldr/api.socket", 60*time.Second)
+	if err != nil {
+		fmt.Printf("ERROR: Test setup failed: %s", err)
+		os.Exit(1)
+	}
+	os.Exit(m.Run())
+}
+
+// POST a new TOML blueprint
+func TestPostTOMLBlueprintV0(t *testing.T) {
+	bp := `
+		name="test-toml-blueprint-v0"
+		description="postTOMLBlueprintV0"
+		version="0.0.1"
+		[[packages]]
+		name="bash"
+		version="*"
+
+		[[modules]]
+		name="util-linux"
+		version="*"
+
+		[[customizations.user]]
+		name="root"
+		password="qweqweqwe"
+		`
+	resp, err := client.PostTOMLBlueprintV0(testState.socket, bp)
+	require.NoError(t, err, "failed with a client error")
+	require.True(t, resp.Status, "POST failed: %#v", resp)
+}
+
+// POST an invalid TOML blueprint
+func TestPostInvalidTOMLBlueprintV0(t *testing.T) {
+	bp := `
+		name="test-invalid-toml-blueprint-v0"
+		version="0.0.1"
+		[package
+		name="bash"
+		version="*"
+		`
+	resp, err := client.PostTOMLBlueprintV0(testState.socket, bp)
+	require.NoError(t, err, "failed with a client error")
+	require.False(t, resp.Status, "did not return an error")
+}
+
+// POST an empty TOML blueprint
+func TestPostEmptyTOMLBlueprintV0(t *testing.T) {
+	resp, err := client.PostTOMLBlueprintV0(testState.socket, "")
+	require.NoError(t, err, "failed with a client error")
+	require.False(t, resp.Status, "did not return an error")
+}
+
+// POST a new JSON blueprint
+func TestPostJSONBlueprintV0(t *testing.T) {
+	bp := `{
+		"name": "test-json-blueprint-v0",
+		"description": "postJSONBlueprintV0",
+		"version": "0.0.1",
+		"packages": [{"name": "bash", "version": "*"}],
+		"modules": [{"name": "util-linux", "version": "*"}],
+		"customizations": {"user": [{"name": "root", "password": "qweqweqwe"}]}
+	}`
+
+	resp, err := client.PostJSONBlueprintV0(testState.socket, bp)
+	require.NoError(t, err, "failed with a client error")
+	require.True(t, resp.Status, "POST failed: %#v", resp)
+}
+
+// POST an invalid JSON blueprint
+func TestPostInvalidJSONBlueprintV0(t *testing.T) {
+	bp := `{
+		"name": "test-invalid-json-blueprint-v0",
+		"version": "0.0.1",
+		"modules": [{"name: "util-linux", "version": "*"}],
+	}`
+
+	resp, err := client.PostJSONBlueprintV0(testState.socket, bp)
+	require.NoError(t, err, "failed with a client error")
+	require.False(t, resp.Status, "did not return an error")
+}
```

Comment:
From what I see the invalid yaml/json scenarios are using payload which isn't valid YAML/JSON. A different scenario is when the payload is a valid file format but the contents inside are invalid, e.g. missing a required field. 

Would also help if you add a pointer/comment as to exactly which part of the string isn't valid. In this case I believe it is the comma on the last line which will choke the JSON parser but that is hard to spot.

Reply:
TOML not YAML :) In these tests I am using either an empty blueprint or one that has a formatting problem. Added comments to the Invalid tests, thanks!